design/ip-address-management-for-networkdata.md

```diff
@@ -0,0 +1,373 @@
+
+### Non-Goals
+
+- provide a DHCP feature or any equivalent level feature.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a user deploying a target Kubernetes cluster, I want to be able to specify
+a non-continuous range of public IP addresses to be used for the nodes static
+allocation as those sparse addresses are all I got from my network admins.
+
+#### Story 2
+
+As a user deploying a target Kubernetes cluster, I want to be able to specify
+the pool of public IP addresses available once, and have any machine deployment
+or kubeadm control plane referencing it use it, effectively sharing it among
+multiple machine deployments.
+
+#### Story 3
+
+As a user deploying a target Kubernetes cluster, I want to keep the allocation
+predictability offered by the current feature, i.e. the ip address is the
+start address + the index of the Metal3Data, when I use a pool for a single
+machine deployment. This is to ensure that the current proposal adds a feature
+without removing any existing feature and is fully backward compatible.
+
+#### Story 4
+
+As a user deploying a target Kubernetes cluster, I want that allocation of
+IP addresses to be preserved during the pivoting operation towards the target
+cluster to prevent future conflicts.
+
+### Implementation Details/Notes/Constraints
+
+- This proposal should be fully backwards compatible and not modify any existing
+  behaviour.
+- When shared among multiple machine deployment, the allocation would be
+  random as much as possible to avoid conflicts and re-use as much as possible.
+- If not needed, this feature should not require the user to modify anything
+  existing in their CRs
+
+### Risks and Mitigations
+
+This adds some complexity to the Metal3 system. Proper documentation could solve
+the issue, and for users not needing this feature, no change would be required.
+
+## Design Details
+
+
+A Metal3IPPool object would be created :
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+kind: Metal3IPPool
+metadata:
+  name: pool-1
+  namespace: default
+  ownerReferences:
+  - apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+    kind: Metal3Cluster
+    name: cluster-1
+  - apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+    kind: Metal3Data
+    name: metal3data-1
+spec:
+  pools:
+    - start: 192.168.0.10
+      end: 192.168.0.15
+      subnet: 192.168.0.0/24
+      gateway: 192.168.0.1
+      prefix: 24
+    - start: 192.168.1.10
+      end: 192.168.1.15
+      subnet: 192.168.1.0/24
+      gateway: 192.168.1.1
+      prefix: 24
+  gateway: 192.168.1.1
+  prefix: 24
+  allocations:
+    "metal3data-10": 192.168.0.9
+    "metal3data-9": 192.168.0.8
+  namePrefix: "provisioning"
+status:
+  lastUpdated: "2020-04-02T06:36:09Z"
+  addresses:
+    "192.168.0.11": "metal3data-1"
+  allocations:
+    "metal3data-1": "pool-1-192-168-0-11"
+```
+
+In the *spec* of *Metal3IPPool*, there would be a *pools* list, that would
```

Comment:
Given that the API is going to be namespace under metal3.io, do we need Metal3 in the names as well?

Reply:
the namespace is actually `infrastructure.cluster.x-k8s.io` so it is not possible unfortunately